Adds docs page for deploying flows via Python (#15140)
PrefectHQ/prefect: docs/3.0rc/deploy/infrastructure-concepts/deploy-via-python.mdx

---
title: Deploy flows with Python
description: Learn how to use the Python SDK to deploy flows to run in work pools.
---

Prefect offers a flexible way to deploy flows to dynamic infrastructure using the Python SDK. 
This approach allows you to target specific work pools and utilize dynamically provisioned infrastructure.

Deploying flows to dynamic infrastructure offers a flexible and programmatic approach to managing your 
Prefect workflows. While `flow.serve` remains a valid and straightforward method for deploying flows to persistent 
infrastructure, there are scenarios where you might need to leverage dynamic, scalable resources. In these cases, 
`flow.deploy` provides a powerful alternative that allows you to target specific work pools and utilize dynamically 
provisioned infrastructure.

## When to consider `flow.deploy` over `flow.serve`

The `flow.serve` method is an excellent choice for many use cases, especially when you have readily available, persistent infrastructure. However, you might want to consider using `flow.deploy` in the following situations:

> 1. Cost optimization: Dynamic infrastructure can help reduce costs by scaling resources up or down based on demand.

Reviewer: I feel like these could be combined into one bullet point on dynamic infrastructure and you could add other bullet points for other situations. For example:

1. Guardrails for other workflow authors in your organization through customized job templates.
2. Scale and redundancy for large, potentially long-running, workloads with K8s

Author: I chose these points because I thought they mapped well onto what users would care about when deciding. I'd prefer not to collapse these ideas since I think they're important to communicate.

2. Resource scarcity: If you have limited persistent infrastructure, dynamic provisioning can help manage resource allocation more efficiently.
3. Varying workloads: For workflows with inconsistent resource needs, dynamic infrastructure can adapt to changing requirements.
4. Cloud-native deployments: When working with cloud providers that offer serverless or auto-scaling options.

By using `flow.deploy`, you can take advantage of Prefect's work pool system, which provides a layer of abstraction between your flows and the underlying infrastructure. This approach offers greater flexibility and efficiency in running your workflows when dynamic infrastructure is needed.

> 

Reviewer: Maybe good to add a note here that workflows that use work pools can alternatively be defined in yaml and they can see the other page to learn more.

Author: I tried not to mention the YAML-based deployment method because I think it would be a distraction on this page.

This page is next to the YAML-based deployment method page in the sidebar, which I think is an appropriate level of discoverability. We can also show links to both methods on pages with links to deployment methods.

Happy to discuss if you have a different view point!

Reviewer: I'm fine with that. I think the improved search functionality should help too.

> Let's explore how to create a deployment using the Python SDK and leverage dynamic infrastructure through work pools.

Reviewer: Maybe note that you're going to use Docker below (maybe because most infrastructure uses it - it's representative) and that there are other options for users depending upon their needs and link to those.

Author: Added in https://github.com/PrefectHQ/prefect/pull/15140/commits/dcea3e76240519e04f85572bf2f5df17035ad627!

        @flow(log_prints=True)
        def my_flow(name: str = "world"):
            print(f"Hello, {name}!")
        ```
    </Step>
    <Step title="Add deployment configuration">
        Add a call to `flow.deploy` to tell Prefect how to deploy your flow.
        ```python example.py
        from prefect import flow

        @flow(log_prints=True)
        def my_flow(name: str = "world"):
            print(f"Hello, {name}!")

        if __name__ == "__main__":
            my_flow.deploy(
                name="my-deployment",
                work_pool_name="my-work-pool",
                image="my-docker-image:dev",
                push=False
            )
        ```
    </Step>
    <Step title="Deploy!">
        Run your script to deploy your flow.
        ```bash
        python example.py
        ```

        Running this script will:

        1. Build a Docker image containing your flow code and dependencies.
        2. Create a deployment associated with the specified work pool and image.
    </Step>
</Steps>

Building a Docker image for our flow allows us to have a consistent environment for our flow to run in. Workers for our work pool will use the image to run our flow.

> In this example, we set `push=False` to skip pushing the image to a registry. This is useful for local development, and you can push your image to a registry like 

Reviewer: ```suggestion
In this example, we set `push=False` to skip pushing the image to a registry. This is useful for local development, and you can push your image to a registry such as  
```

> Docker Hub for more production environments.

Reviewer: ```suggestion
Docker Hub in a production environment.
```

<Note>
**Where's the Dockerfile?**

In the above example, we didn't specify a Dockerfile. By default, Prefect will generate a Dockerfile for us that copies the flow code into an image and installs 
any additional dependencies.

If you want to write and use your own Dockerfile, you can do so by passing a `dockerfile` parameter to `flow.deploy`.
</Note>

> ### Triggering a run

Reviewer: ```suggestion
### Trigger a run
```

Now that we have our flow deployed, we can trigger a run via either the Prefect CLI or UI.

First, we need to start a worker to run our flow:

```bash
prefect worker start --pool my-work-pool
```

Then, we can trigger a run of our flow using the Prefect CLI:

```bash
prefect deployment run 'my-flow/my-deployment'
```

> After a few seconds, you should see logs from your worker showing that the flow run has started and see the state update in the UI!

Reviewer: ```suggestion
After a few seconds, you should see logs from your worker showing that the flow run has started and see the state update in the UI.
```

        - Setting up various recurring schedules
        - Implementing complex scheduling logic
        - Applying timezone offsets to schedules

    <Expandable title="Example">
    ```python schedules.py
    from datetime import datetime, timedelta
    from prefect import flow
    from prefect.client.schemas.schedules import IntervalSchedule

    @flow(log_prints=True)
    def my_flow(name: str = "world"):
        print(f"Hello, {name}!")

    if __name__ == "__main__":
        my_flow.deploy(
            name="my-deployment",
            work_pool_name="my-work-pool",
            image="my-docker-image:dev",
            push=False,
            # Run every 10 minutes starting from January 1, 2023
            # at 00:00 Central Time
            schedules=[
                IntervalSchedule(
                    interval=timedelta(minutes=10),
                    anchor_date=datetime(2023, 1, 1, 0, 0),
                    timezone="America/Chicago"
                )
            ]
        )
    ```
    </Expandable>

    Learn more about schedules [here](/3.0rc/automate/add-schedules).

## Use remote code storage

> In addition to using Docker images, Prefect also supports deploying your code using remote storage. This approach allows you to 

Reviewer: ```suggestion
In addition to storing your  code in a Docker image, Prefect also supports deploying your code to remote storage. This approach allows you to 
```

store your flow code in a remote location, such as a Git repository or cloud storage service.

> Using remote storage for your code has several advantages:

Reviewer 2: loosely held nit:

wondering if we could add some statement to the effect of "github is where your code lives anyways" or something, bc often I notice people feel arbitrarily pigeonholed and forget github is an option

Author: Yeah, I can add something about why a remote git repository works well for code storage.

1. Faster iterations: You can update your flow code without rebuilding Docker images.
2. Reduced storage requirements: You don't need to store large Docker images for each code version.
3. Flexibility: You can use different storage backends based on your needs and infrastructure.

To deploy using remote storage, you'll need to specify where your code is stored by using `flow.from_source` to first load your flow code from a remote location.

Here's an example of loading a flow from a Git repository and deploying it:

```python git-deploy.py
from prefect import flow

if __name__ == "__main__":
    flow.from_source(
        source="https://github.com/username/repository.git",
        entrypoint="path/to/your/flow.py:your_flow_function"
    ).deploy(
        name="my-deployment",
        work_pool_name="my-work-pool",
    )
```

The `source` parameter can be a variety of remote storage options including:

- Git repositories
- S3 buckets
- Google Cloud Storage buckets

> - Azure Blob Storage

Reviewer 2: could we get an example of what each one of these looks like? (e.g. s3://my-bucket-etc)

Reviewer: FWIW, we've got some of this in the flow code storage page, as I recall.

```python default-parameters.py
from prefect import flow

@flow
def my_flow(name: str = "world"):
    print(f"Hello, {name}!")

if __name__ == "__main__":
    my_flow.deploy(
        name="my-deployment",
        work_pool_name="my-work-pool",
        # Will print "Hello, Marvin!" by default instead of "Hello, world!"
        parameters={"name": "Marvin"},
        image="my-docker-image:dev",
        push=False,
    )
```

Note these parameters can still be overridden on a per-deployment basis.

## Set job variables

You can set default job variables for a deployment using the `job_variables` keyword argument in `flow.deploy`. 
The job variables provided will override the values set on the work pool.

```python job-variables.py
import os
from prefect import flow

@flow
def my_flow():
    name = os.getenv("NAME", "world")
    print(f"Hello, {name}!")

if __name__ == "__main__":
    my_flow.deploy(
        name="my-deployment",
        work_pool_name="my-work-pool",
        # Will print "Hello, Marvin!" by default instead of "Hello, world!"
        job_variables={"env": {"NAME": "Marvin"}},
        image="my-docker-image:dev",
        push=False,
    )
```

> Job variables can be used to customize environment variables, resources limits, and other infrastructure options

Reviewer 2: ```suggestion
[Job variables](/3.0rc/deploy/infrastructure-concepts/customize#override-job-variables) can be used to customize environment variables, resources limits, and other infrastructure options
```

Author: There's a link to that page at the end of this section. I like it at the end so that the flow is you read through this and then go read more if you want/need to.

>  allowing fine-grained control over your infrastructure on a per-deployment basis.

Reviewer 2: ```suggestion
 allowing fine-grained control over your infrastructure on a per-deployment or per-flow-run basis.
```
